analyze/aps-cycle.c

```diff
@@ -711,14 +730,196 @@ static void add_up_down_attributes(STATE *s, bool direction)
+{
+  void *current = decl;
+  Declaration current_decl;
+  while (current != NULL && (current = tnode_parent(current)) != NULL)
+  {
+    switch (ABSTRACT_APS_tnode_phylum(current))
+    {
+    case KEYDeclaration:
+      current_decl = (Declaration)current;
+      switch (Declaration_KEY(current_decl))
+      {
+      case KEYsome_function_decl:
+        *func = current_decl;
+        return some_function_decl_result(current_decl) == decl;
+      default:
+        break;
+      }
+      default:
+        break;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * This function determines whether attribute instance should be considered for circularity check
+ * @param instance attribute instance
+ * @return true if instance not is not: If, Match and some formal
+ * @return false everything else
+ */
+static bool applicable_for_circularity_check(INSTANCE *instance)
+{
+  // If and Match statements can show up in a cycle but are never declared circular or non-circular
+  if (if_rule_p(instance->fibered_attr.attr))
+    return false;
+
+  Declaration node = instance->fibered_attr.attr;
+  Declaration func = NULL;
+
+  // The result in a function/procedure may show up in a cycle but are never declared circular or non-circular
+  if (is_inside_some_function(node, &func) && some_function_decl_result(func) == node)
+    return false;
+
+  if (instance->fibered_attr.fiber != NULL)
+    return false;
+
+  // Formals may show up in a cycle as well
+  switch (ABSTRACT_APS_tnode_phylum(node))
+  {
+  case KEYDeclaration:
+  {
+    switch (Declaration_KEY(node))
+    {
+    case KEYformal:
+      return false;
+    default:
+      break;
+    }
+  }
+  default:
+    break;
+  }
+
+  return true;
+}
+
+/**
+ * Ensure that attributes that participate in a cycle are defined circular
+ * @param s Analysis state
+ */
+static void assert_circular_declaration(STATE* s) {
+  int i, j, k;
+
+  // Forall phylum in the phylum_graph
+  for (i = 0; i < s->phyla.length; i++) {
+    PHY_GRAPH* phy = &s->phy_graphs[i];
+    int n = phy->instances.length;
+    int phylum_index = phylum_instance_start[i];
+    INSTANCE* array = phy->instances.array;
+
+    for (j = 0; j < n; j++) {
+      INSTANCE* instance = &array[j];
+      Declaration node = instance->fibered_attr.attr;
+
+      if (!applicable_for_circularity_check(instance))
+        continue;
+
+      bool any_cycle = false;
+      bool declared_circular = instance_circular(instance);
+
+      char instance_to_str[BUFFER_SIZE];
+      FILE* f = fmemopen(instance_to_str, sizeof(instance_to_str), "w");
+      print_instance(instance, f);
+      fclose(f);
+
+      for (k = 0; k < num_instances; k++) {
+        if (parent_index[k] == k) {
+          if (parent_index[phylum_index + j] == k) {
+            if (declared_circular) {
+              any_cycle = true;
+            } else {
+              aps_error(node,
+                        "Phylum graph (%s) instance (%s) involves in a cycle "
```

Comment:
why "Phylum graph (%s) ..."
Why should the user be told about the phylum graph?
Perhaps the phylum should be listed, but even then, isn't the instance enough information?

Reply:
Fixed.